File: lib/src/androidTest/java/io/github/lucf15/tiffrenderer/TiffRendererRenderTest.java
/*
 * Copyright 2026 lucf15
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *      http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package io.github.lucf15.tiffrenderer;

import static org.junit.Assert.assertEquals;
import static org.junit.Assert.assertFalse;
import static org.junit.Assert.assertThrows;

import android.graphics.Bitmap;
import android.graphics.Color;
import android.graphics.Matrix;
import android.graphics.Rect;
import android.os.ParcelFileDescriptor;

import androidx.test.ext.junit.runners.AndroidJUnit4;

import org.junit.Test;
import org.junit.runner.RunWith;

import java.io.IOException;

/**
 * {@link TiffRenderer.Page#render}'s argument validation and pixel-level correctness. Uses
 * {@code single_page_rgb.tif} (32x24, solid RGB) throughout -- a flat color keeps pixel
 * assertions exact regardless of nearest-neighbor vs. bilinear resampling.
 */
@RunWith(AndroidJUnit4.class)
public class TiffRendererRenderTest {

    private static final int PAGE_WIDTH = 32;
    private static final int PAGE_HEIGHT = 24;
    private static final int PAGE_COLOR = Color.rgb(128, 64, 200);

    private ParcelFileDescriptor openSinglePage() throws IOException {
        return TestAssets.open("single_page_rgb.tif");
    }

    @Test
    public void render_nullDestination_throwsNullPointerException() throws IOException {
        try (ParcelFileDescriptor pfd = openSinglePage();
                TiffRenderer renderer = new TiffRenderer(pfd)) {
            TiffRenderer.Page page = renderer.openPage(0);
            assertThrows(NullPointerException.class,
                    () -> page.render(null, null, null, TiffRenderer.Page.RENDER_MODE_FOR_DISPLAY));
            page.close();
        }
    }

    @Test
    public void render_wrongBitmapConfig_throwsIllegalArgumentException() throws IOException {
        try (ParcelFileDescriptor pfd = openSinglePage();
                TiffRenderer renderer = new TiffRenderer(pfd)) {
            TiffRenderer.Page page = renderer.openPage(0);
            Bitmap bitmap = Bitmap.createBitmap(PAGE_WIDTH, PAGE_HEIGHT, Bitmap.Config.RGB_565);
            assertThrows(IllegalArgumentException.class, () -> page.render(
                    bitmap, null, null, TiffRenderer.Page.RENDER_MODE_FOR_DISPLAY));
            page.close();
        }
    }

    @Test
    public void render_destClipExceedsBitmapBounds_throwsIllegalArgumentException()
            throws IOException {
        try (ParcelFileDescriptor pfd = openSinglePage();
                TiffRenderer renderer = new TiffRenderer(pfd)) {
            TiffRenderer.Page page = renderer.openPage(0);
            Bitmap bitmap = Bitmap.createBitmap(10, 10, Bitmap.Config.ARGB_8888);
            Rect tooWide = new Rect(0, 0, 20, 5);
            assertThrows(IllegalArgumentException.class, () -> page.render(
                    bitmap, tooWide, null, TiffRenderer.Page.RENDER_MODE_FOR_DISPLAY));
            page.close();
        }
    }

    @Test
    public void render_destClipDegenerate_throwsIllegalArgumentException() throws IOException {
        try (ParcelFileDescriptor pfd = openSinglePage();
                TiffRenderer renderer = new TiffRenderer(pfd)) {
            TiffRenderer.Page page = renderer.openPage(0);
            Bitmap bitmap = Bitmap.createBitmap(10, 10, Bitmap.Config.ARGB_8888);
            Rect empty = new Rect(5, 5, 5, 8); // left == right
            assertThrows(IllegalArgumentException.class, () -> page.render(
                    bitmap, empty, null, TiffRenderer.Page.RENDER_MODE_FOR_DISPLAY));
            page.close();
        }
    }

    @Test
    public void render_nonAffineTransform_throwsIllegalArgumentException() throws IOException {
        try (ParcelFileDescriptor pfd = openSinglePage();
                TiffRenderer renderer = new TiffRenderer(pfd)) {
            TiffRenderer.Page page = renderer.openPage(0);
            Bitmap bitmap = Bitmap.createBitmap(PAGE_WIDTH, PAGE_HEIGHT, Bitmap.Config.ARGB_8888);
            Matrix nonAffine = new Matrix();
            float[] values = new float[9];
            nonAffine.getValues(values);
            values[6] = 0.001f; // MPERSP_0 -- nonzero makes Matrix#isAffine() false
            nonAffine.setValues(values);
            assertFalse(nonAffine.isAffine());
            assertThrows(IllegalArgumentException.class, () -> page.render(
                    bitmap, null, nonAffine, TiffRenderer.Page.RENDER_MODE_FOR_DISPLAY));
            page.close();
        }
    }

    @Test
    public void render_invalidRenderMode_throwsIllegalArgumentException() throws IOException {
        try (ParcelFileDescriptor pfd = openSinglePage();
                TiffRenderer renderer = new TiffRenderer(pfd)) {
            TiffRenderer.Page page = renderer.openPage(0);
            Bitmap bitmap = Bitmap.createBitmap(PAGE_WIDTH, PAGE_HEIGHT, Bitmap.Config.ARGB_8888);
            assertThrows(IllegalArgumentException.class,
                    () -> page.render(bitmap, null, null, 99));
            page.close();
        }
    }

    @Test
    public void render_afterPageClosed_throwsIllegalStateException() throws IOException {
        try (ParcelFileDescriptor pfd = openSinglePage();
                TiffRenderer renderer = new TiffRenderer(pfd)) {
            TiffRenderer.Page page = renderer.openPage(0);
            page.close();
            Bitmap bitmap = Bitmap.createBitmap(PAGE_WIDTH, PAGE_HEIGHT, Bitmap.Config.ARGB_8888);
            assertThrows(IllegalStateException.class, () -> page.render(
                    bitmap, null, null, TiffRenderer.Page.RENDER_MODE_FOR_DISPLAY));
        }
    }

    @Test
    public void render_defaultTransform_fillsBitmapWithExactPageColor() throws IOException {
        try (ParcelFileDescriptor pfd = openSinglePage();
                TiffRenderer renderer = new TiffRenderer(pfd)) {
            TiffRenderer.Page page = renderer.openPage(0);
            Bitmap bitmap = Bitmap.createBitmap(PAGE_WIDTH, PAGE_HEIGHT, Bitmap.Config.ARGB_8888);
            page.render(bitmap, null, null, TiffRenderer.Page.RENDER_MODE_FOR_DISPLAY);

            assertEquals(PAGE_COLOR, bitmap.getPixel(0, 0));
            assertEquals(PAGE_COLOR, bitmap.getPixel(PAGE_WIDTH - 1, PAGE_HEIGHT - 1));
            assertEquals(255, Color.alpha(bitmap.getPixel(PAGE_WIDTH / 2, PAGE_HEIGHT / 2)));
            page.close();
        }
    }

    @Test
    public void render_nearestVsBilinearRenderMode_bothExactForFlatColor() throws IOException {
        try (ParcelFileDescriptor pfd = openSinglePage();
                TiffRenderer renderer = new TiffRenderer(pfd)) {
            TiffRenderer.Page page = renderer.openPage(0);
            Bitmap bitmap = Bitmap.createBitmap(PAGE_WIDTH, PAGE_HEIGHT, Bitmap.Config.ARGB_8888);
            page.render(bitmap, null, null, TiffRenderer.Page.RENDER_MODE_FOR_PRINT);

            assertEquals(PAGE_COLOR, bitmap.getPixel(0, 0));
            assertEquals(PAGE_COLOR, bitmap.getPixel(PAGE_WIDTH - 1, PAGE_HEIGHT - 1));
            page.close();
        }
    }

    @Test
    public void render_withDestClip_leavesOutsideClipUntouched() throws IOException {
        try (ParcelFileDescriptor pfd = openSinglePage();
                TiffRenderer renderer = new TiffRenderer(pfd)) {
            TiffRenderer.Page page = renderer.openPage(0);
            Bitmap bitmap = Bitmap.createBitmap(64, 48, Bitmap.Config.ARGB_8888);
            int sentinel = Color.YELLOW;
            bitmap.eraseColor(sentinel);

            Rect clip = new Rect(10, 10, 20, 18);
            page.render(bitmap, clip, null, TiffRenderer.Page.RENDER_MODE_FOR_DISPLAY);

            assertEquals(PAGE_COLOR, bitmap.getPixel(15, 14));
            assertEquals(sentinel, bitmap.getPixel(5, 5));
            assertEquals(sentinel, bitmap.getPixel(9, 14));
            assertEquals(sentinel, bitmap.getPixel(20, 14));
            assertEquals(sentinel, bitmap.getPixel(15, 9));
            assertEquals(sentinel, bitmap.getPixel(15, 18));
            page.close();
        }
    }

    @Test
    public void render_withCustomTranslateTransform_offsetsContentAsSpecified() throws IOException {
        try (ParcelFileDescriptor pfd = openSinglePage();
                TiffRenderer renderer = new TiffRenderer(pfd)) {
            TiffRenderer.Page page = renderer.openPage(0);
            Bitmap bitmap = Bitmap.createBitmap(50, 40, Bitmap.Config.ARGB_8888);
            int sentinel = Color.BLUE;
            bitmap.eraseColor(sentinel);

            Matrix identityTranslated = new Matrix();
            identityTranslated.postTranslate(5, 5);
            page.render(bitmap, null, identityTranslated, TiffRenderer.Page.RENDER_MODE_FOR_DISPLAY);

            assertEquals(sentinel, bitmap.getPixel(0, 0));
            assertEquals(PAGE_COLOR, bitmap.getPixel(5, 5));
            assertEquals(PAGE_COLOR, bitmap.getPixel(5 + PAGE_WIDTH - 1, 5 + PAGE_HEIGHT - 1));
            page.close();
        }
    }
}

File: lib/core/src/androidDeviceTest/kotlin/io/github/lucf15/tiffrenderer/TiffRendererRenderTest.kt
package io.github.lucf15.tiffrenderer

import android.graphics.Bitmap
import android.graphics.Color
import android.graphics.Matrix
import android.graphics.Rect
import org.junit.Assert.assertEquals
import org.junit.Assert.assertThrows
import org.junit.Test

/**
 * [TiffPage.render]'s argument validation and pixel-level correctness, using a flat-color fixture
 * so pixel assertions stay exact. Three cases have no equivalent here: the common Kotlin types
 * [TiffPage.render] takes are structurally stronger than a raw Bitmap/Rect/Matrix/int quartet, so
 * the invalid states those cases would need can't be constructed at all:
 *  - null destination: `destination` is a non-null [TiffBitmap], so passing `null` is a compile
 *    error, not a runtime check.
 *  - non-affine transform: [TiffTransform] only ever stores the 6 affine components; there's no
 *    way to construct a non-affine one to pass.
 *  - invalid render mode: [TiffRenderMode] is a Kotlin enum; there's no "invalid" ordinal.
 * Two more cases (wrong bitmap config, degenerate clip) survive but moved to where the validation
 * now actually happens: [TiffBitmap]'s and [TiffRect]'s own constructors (see their `init`
 * blocks), since both fail fast well before a `render()` call.
 */
class TiffRendererRenderTest {

    companion object {
        private const val PAGE_WIDTH = 32
        private const val PAGE_HEIGHT = 24
        private val PAGE_COLOR = Color.rgb(128, 64, 200)
    }

    private fun openSinglePage() =
        TiffRenderer(TiffSource.fromParcelFileDescriptor(TestFixtures.open("single_page_rgb.tif")))

    @Test
    fun tiffBitmap_wrongBitmapConfig_throwsIllegalArgumentException() {
        val bitmap = Bitmap.createBitmap(PAGE_WIDTH, PAGE_HEIGHT, Bitmap.Config.RGB_565)
        assertThrows(IllegalArgumentException::class.java) { TiffBitmap(bitmap) }
    }

    @Test
    fun render_destClipExceedsBitmapBounds_throwsIllegalArgumentException() {
        openSinglePage().use { renderer ->
            val page = renderer.openPage(0)
            val bitmap = TiffBitmap(Bitmap.createBitmap(10, 10, Bitmap.Config.ARGB_8888))
            val tooWide = TiffRect(0, 0, 20, 5)
            assertThrows(IllegalArgumentException::class.java) {
                page.render(bitmap, tooWide, null, TiffRenderMode.FOR_DISPLAY)
            }
            page.close()
        }
    }

    @Test
    fun tiffRect_degenerate_throwsIllegalArgumentException() {
        assertThrows(IllegalArgumentException::class.java) { TiffRect(5, 5, 5, 8) } // left == right
    }

    @Test
    fun render_afterPageClosed_throwsIllegalStateException() {
        openSinglePage().use { renderer ->
            val page = renderer.openPage(0)
            page.close()
            val bitmap = TiffBitmap(Bitmap.createBitmap(PAGE_WIDTH, PAGE_HEIGHT, Bitmap.Config.ARGB_8888))
            assertThrows(IllegalStateException::class.java) {
                page.render(bitmap, null, null, TiffRenderMode.FOR_DISPLAY)
            }
        }
    }

    @Test
    fun render_defaultTransform_fillsBitmapWithExactPageColor() {
        openSinglePage().use { renderer ->
            val page = renderer.openPage(0)
            val bitmap = Bitmap.createBitmap(PAGE_WIDTH, PAGE_HEIGHT, Bitmap.Config.ARGB_8888)
            page.render(TiffBitmap(bitmap), renderMode = TiffRenderMode.FOR_DISPLAY)

            assertEquals(PAGE_COLOR, bitmap.getPixel(0, 0))
            assertEquals(PAGE_COLOR, bitmap.getPixel(PAGE_WIDTH - 1, PAGE_HEIGHT - 1))
            assertEquals(255, Color.alpha(bitmap.getPixel(PAGE_WIDTH / 2, PAGE_HEIGHT / 2)))
            page.close()
        }
    }

    @Test
    fun render_nearestVsBilinearRenderMode_bothExactForFlatColor() {
        openSinglePage().use { renderer ->
            val page = renderer.openPage(0)
            val bitmap = Bitmap.createBitmap(PAGE_WIDTH, PAGE_HEIGHT, Bitmap.Config.ARGB_8888)
            page.render(TiffBitmap(bitmap), renderMode = TiffRenderMode.FOR_PRINT)

            assertEquals(PAGE_COLOR, bitmap.getPixel(0, 0))
            assertEquals(PAGE_COLOR, bitmap.getPixel(PAGE_WIDTH - 1, PAGE_HEIGHT - 1))
            page.close()
        }
    }

    @Test
    fun render_withDestClip_leavesOutsideClipUntouched() {
        openSinglePage().use { renderer ->
            val page = renderer.openPage(0)
            val bitmap = Bitmap.createBitmap(64, 48, Bitmap.Config.ARGB_8888)
            val sentinel = Color.YELLOW
            bitmap.eraseColor(sentinel)

            val clip = TiffRect(10, 10, 20, 18)
            page.render(TiffBitmap(bitmap), clip, null, TiffRenderMode.FOR_DISPLAY)

            assertEquals(PAGE_COLOR, bitmap.getPixel(15, 14))
            assertEquals(sentinel, bitmap.getPixel(5, 5))
            assertEquals(sentinel, bitmap.getPixel(9, 14))
            assertEquals(sentinel, bitmap.getPixel(20, 14))
            assertEquals(sentinel, bitmap.getPixel(15, 9))
            assertEquals(sentinel, bitmap.getPixel(15, 18))
            page.close()
        }
    }

    @Test
    fun render_withCustomTranslateTransform_offsetsContentAsSpecified() {
        openSinglePage().use { renderer ->
            val page = renderer.openPage(0)
            val bitmap = Bitmap.createBitmap(50, 40, Bitmap.Config.ARGB_8888)
            val sentinel = Color.BLUE
            bitmap.eraseColor(sentinel)

            // TiffTransform is [mxx, mxy, mtx, myx, myy, mty]: identity scale, translate by (5,5).
            val translated = TiffTransform(floatArrayOf(1f, 0f, 5f, 0f, 1f, 5f))
            page.render(TiffBitmap(bitmap), null, translated, TiffRenderMode.FOR_DISPLAY)

            assertEquals(sentinel, bitmap.getPixel(0, 0))
            assertEquals(PAGE_COLOR, bitmap.getPixel(5, 5))
            assertEquals(PAGE_COLOR, bitmap.getPixel(5 + PAGE_WIDTH - 1, 5 + PAGE_HEIGHT - 1))
            page.close()
        }
    }

    @Test
    fun render_androidNativeOverload_matchesCommonTypesResult() {
        openSinglePage().use { renderer ->
            val page = renderer.openPage(0)
            val bitmap = Bitmap.createBitmap(50, 40, Bitmap.Config.ARGB_8888)
            val sentinel = Color.BLUE
            bitmap.eraseColor(sentinel)

            val matrix = Matrix().apply { setTranslate(5f, 5f) }
            page.render(bitmap, null, matrix, TiffRenderMode.FOR_DISPLAY)

            assertEquals(sentinel, bitmap.getPixel(0, 0))
            assertEquals(PAGE_COLOR, bitmap.getPixel(5, 5))
            assertEquals(PAGE_COLOR, bitmap.getPixel(5 + PAGE_WIDTH - 1, 5 + PAGE_HEIGHT - 1))
            page.close()
        }
    }

    @Test
    fun render_androidNativeOverload_destClipMatchesTiffRectResult() {
        openSinglePage().use { renderer ->
            val page = renderer.openPage(0)
            val bitmap = Bitmap.createBitmap(64, 48, Bitmap.Config.ARGB_8888)
            val sentinel = Color.YELLOW
            bitmap.eraseColor(sentinel)

            page.render(bitmap, Rect(10, 10, 20, 18), null, TiffRenderMode.FOR_DISPLAY)

            assertEquals(PAGE_COLOR, bitmap.getPixel(15, 14))
            assertEquals(sentinel, bitmap.getPixel(5, 5))
            page.close()
        }
    }

    @Test
    fun matrixToTiffTransform_nonAffineMatrix_throwsIllegalArgumentException() {
        val matrix = Matrix().apply {
            setValues(floatArrayOf(1f, 0f, 0f, 0f, 1f, 0f, 0.1f, 0f, 1f))
        }
        assertThrows(IllegalArgumentException::class.java) { matrix.toTiffTransform() }
    }

    @Test
    fun tiffRenderer_androidNativeOverload_opensPfdDirectly() {
        TiffRenderer(TestFixtures.open("single_page_rgb.tif")).use { renderer ->
            assertEquals(1, renderer.pageCount)
        }
    }
}
